### DISCUSSION TITLE
NetworkError when attempting to fetch resource in Firefox

### DISCUSSION THREAD

**Discussion by Li-amK** [community/community#17126] [Codespaces]
So i recently got Acess to the Codespaces Beta and naturally was super excited to get started.

Everything works smoothly in Chromium based Browsers, but i just wanted to let you know that Firefox produces the folowing:
![grafik](https://user-images.githubusercontent.com/65390745/169711463-1cb3df2d-315a-4deb-aa00-a87f5fa5c77f.png)

## Other Infos
**Firefox Version:** 100.0.2 (64-Bit) (All Extensions Disabled)
**OS:** Windows 10 Pro Version 21H2 (Build 19044.1706)

**Comment by legomushroom** [comment]
@Li-amK thanks for reporting! Mind sending me devtools logs from your browser? legomushroom@github.com

**Reply by Li-amK** [reply]
@legomushroom sure thing! Check your Inbox.

Good to hear I'm not alone with this Issue

**Comment by zurgeg** [comment]
Ah! So it's not just me! I've opened a ticket with GitHub support, I have this same problem. I hope that they will fix this soon!

@legomushroom The Firefox devtools say this (for me at least, OP seems to have the same problem though):
 ```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://api.github.com/user/codespaces/foobar-foobar-codespaces-is-broken-deadbeef?internal=true&refresh=true&t=(very big number). (Reason: CORS request did not succeed). Status code: (null)
Failed to load ‘https://api.github.com/user/codespaces/foobar-foobar-codespaces-is-broken-deadbeef?internal=true&refresh=true&t=(very big number)’. A ServiceWorker passed a promise to FetchEvent.respondWith() that rejected with ‘TypeError: NetworkError when attempting to fetch resource.’
```

Take note of **`The Same Origin Policy disallows reading the remote resource`**. I decided to try and **turn off CORS**. I found [this post](https://stackoverflow.com/questions/17711924/disable-cross-domain-web-security-in-firefox) on Stack Overflow, which lead me to [this](https://addons.mozilla.org/en-US/firefox/addon/cors-everywhere/) extension. Sure enough, I install it, turn it on. Refresh the page, and sure enough, it worked.

Maybe make an `api.github.dev`?

**Comment by legomushroom** [comment]
Thanks for the logs, looking at them now 👀

**Reply by legomushroom** [reply]
@Li-amK @zurgeg are you using some Ad-block/privacy extension in your firefox? I was not repro this so far, so an extension wiping out the `Access-Control-Allow-Origin` might explain the issue with CORS 🤔  Mind trying with all extensions disabled?

**Reply by zurgeg** [reply]
Interesting. I use UBO and Firefox's default tracking protection, but disabling both did nothing

**Reply by zurgeg** [reply]
Running in Firefox Troubleshoot mode didn't fix the problem.

**Reply by legomushroom** [reply]
@zurgeg mind sharing your OS and Firefox version/details? Will try to setup the same environment as you ☺️

**Reply by zurgeg** [reply]
Sure!

OS: Win 11
Firefox: Standard 100.0.2 (64-bit)

**Reply by Li-amK** [reply]
I do use the DuckDuckGo plugin and firefox’ own privacy settings, although I already tried disabling all extensions before reporting the Bug.

**Reply by zurgeg** [reply]
@legomushroom FYI, the customer support agent I was working with sent you my devconsole logs. Can you check if we both have the same problem? I'm pretty sure of it but it would be nice to get some confirmation!

**Reply by legomushroom** [reply]
@zurgeg yes I think you have the same issue.

**Reply by legomushroom** [reply]
@Li-amK @zurgeg mind trying this:

Tried 4 different machines/OSes and Codespaces with different age and no repro so far 🥸 

Mind trying this? ☺️ 

1. Unregister ServiceWorker:

- open `about:serviceworkers`
- click on `Unregister` button for the SW of the `*.github.dev` domain
- reload the page and see if it is able to connect now

If that does not help, 

2. Disable service workers completely:

- go to `about:config` and search for `dom.serviceWorkers.enabled` and change it to `false`
- unregister currently running SW as described in `1.`
- reload the page and see if it is able to connect now

If that does not fix the issue, try

3. Clearing out all storage for the `*.github.dev` domain:

- open devtools, go to `Storage` tab
- for each item in the sidebar on the left (e.g. `Cache Storage`, `Cookies` etc.), right-click on the `*.github.dev` doman -> `Delete All`
- reload the page and see if it is able to connect now

Thanks for bearing with me 🙏

**Reply by zurgeg** [reply]
Sorry, none of these worked :(

Can you tell me the setup on those machines so I can try to fix this?

**Reply by zurgeg** [reply]
A [CORS tester](https://cors-test.codehappy.dev/?url=https%3A%2F%2Fapi.github.com%2F&method=get) suggests that `Allow-Access-Control-Origin` (or whatever the header is, I forgot ;)) is not set up on `api.github.com` 🤨

**Reply by legomushroom** [reply]
@zurgeg yeah that header definitely set otherwise it won't work in all browsers.

I tried latest Firefox with default settings on:

1. `12.3.1` Mac OSx Monterey
2. Windows 10
3. `15.1` iOS on IPad Pro
4. `18.04` Ubuntu on ARM11 machine (Raspberry PI 4)

Also tried running, fresh shutdown, 1 month old shutdown and 3 months old shutdown Codespaces on those machines. 

I suspect that the request is failing for some other reason (e.g. 401 403 etc). In those cases the header is not present so browser surfaces it as CORS issue 🤔 

Mind doing this:

- go to `about:debugging#/runtime/this-firefox` and find currently running SW for the `*.github.dev` domain.
- click on `Inspect` button to open devtools in the SW context
- reload the Codespace page and grab all the logs from there
- send them over to me at legomushroom@github.com

Are you up for a zoom call to debug this on your machine?

**Reply by zurgeg** [reply]
Well, just before you said that, I was inspecting, and I saw this upon attempting to replicate the request.
```json
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/reference/codespaces#get-a-codespace-for-the-authenticated-user"
}
```
But even then, the Access Control header is there...

I'll try and send over logs ASAP.

**Reply by zurgeg** [reply]
Alright, @legomushroom, I copied the fetch request from the (failed) CORS API request. I then changed the request type from `cors` to `no-cors`, which caused the request to go through. Can you check the code responsible for accessing the API and see if you are calling `fetch`?

**Reply by Li-amK** [reply]
Sooooo none of the suggested fixes worked @legomushroom 
I have Emailed you my logs from Inspecting the Service Worker too.

**Reply by zurgeg** [reply]
@Li-amK You can try using the [CORS](https://addons.mozilla.org/en-US/firefox/addon/cors-everywhere/) extension I mentioned before until they fix the problem. It is a shame that we can't do it by default yet :(

**Reply by Li-amK** [reply]
@zurgeg i figured i'll be doing exactly that. thanks for the hotfix

**Reply by legomushroom** [reply]
@zurgeg we use no `mode` settings in this request. `mode: "no-cors"` won't work for us, since we need response body to be present, but `no-cors` would result in opaque requests 🥸 

@zureg @Li-amK what timezone(GMT) are you in? Would be great to have a screenshare call to debug it on your machines? I'm in Seattle area, `GMT-7` ☺️

**Reply by Li-amK** [reply]
@legomushroom sure, we can set up a screenshare. I‘m in Germany, `GMT+2` though.

**Reply by transcental** [reply]
I've got the same issue as well with Firefox Dev Ed 102.0b3. The CORS extension partially helps but the icons don't load. I think that's a VSC problem though. (https://github.com/microsoft/vscode/issues/151293)
